src/Database/LSMTree/Internal/Merge.hs

```diff
@@ -61,33 +74,74 @@ new ::
   -> Mappend
   -> Run.RunFsPaths
   -> V.Vector (Run IO (FS.Handle h))
-  -> IO (Maybe (Merge IO (FS.Handle h)))
+  -> IO (Maybe (Merge IO h))
 new fs hbio mergeCaching alloc mergeLevel mergeMappend targetPaths runs = do
     -- no offset, no write buffer
     mreaders <- Readers.new fs hbio Readers.NoOffsetKey Nothing runs
     for mreaders $ \mergeReaders -> do
       -- calculate upper bounds based on input runs
       let numEntries = coerce (sum @V.Vector @Int) (fmap Run.runNumEntries runs)
       mergeBuilder <- Builder.new fs targetPaths numEntries alloc
-      return Merge {..}
+      mergeLastStepResult <- newMutVar $! MergeInProgress
+      return Merge {
+          mergeHasFS = fs
+        , mergeHasBlockIO = hbio
+        , ..
+        }
+
```

Comment:
You could name the first two function arguments `mergeHasFS` and `mergeHasBlockIO` to omit the explicit record assignment. Your call.